Upgrade brotli to 0.6.0

aidanhs opened the pull request:
As a side effect, the bindings have become much simplified because they don't need to contain all the internal details of the decoder struct any more.

Some thoughts before merging

1. `BrotliDecompressedSize` is [gone](https://github.com/google/brotli/commit/f20b3eeb2f401fc4b87488ca22ca32729a6dec20#diff-1721da9a2c189d371911e516cdd0b5b1L1302) with no replacement. This means I've removed the `decompressed_size` method, which doesn't make me too sad since the comment seemed to basicaly say "this isn't reliable". On the other hand, it's backwards incompatible. If we really wanted to preserve this rather than bump version, we could ask the brotli project if they plan to add anything back.
2. I've removed some commented out methods using APIs that were deprecated and are now removed.
3. The API we're now making use of is deprecated (as revealed by `ctest`, which I've not come across before and is nifty), so I've disabled deprecated warnings for now. I've only done this on non-msvc because a look at `ctest` seems to indicate that this warning is [disabled by default](https://github.com/alexcrichton/ctest/blob/0b7f7b707add9d10c85bd3911030f7a739c349fa/src/lib.rs#L567) for msvc for some reason (been there since the initial commit)? (I've not actually tested msvc)
4. I've added a few TODOs for using the newer APIs (`BrotliEncoderCompressStream` function for compression, more compression parameters, human readable errors when decoding) that didn't seem particularly crucial for this upgrade.

aidanhs commented:
Regarding "The API we're now making use of is deprecated" - this is the 'low level API' and was [marked as deprecated](https://github.com/google/brotli/commit/0a63f99db9eeb8a15dc3941b96802141b1993912#diff-458c1c05f868f84a3aefbc0f34575ee1R95) with the recommended new route appearing to be the high level API.

Moving to use the higher-level API effectively brings the `Compress` struct in line with `Decompress`, i.e. no more `copy_input` and a change in signature of the `compress` function. This is (also) a breaking change so is probably worth considering as part of this PR.

If this was a pure rust project I'd be pretty gung-ho about all of this, but multiple versions of brotli can't be linked into the same application which makes it tricky. Any thoughts on best ways to do this?

alexcrichton commented:
Thanks for this! It looks like a number of breaking changes have happened at the actual library itself, so I'm totally fine just releasing new major versions of this and the sys crate. You're right that `links` may cause problems, but this isn't necessarily a hugely shared dep so it may not be that hard to upgrade?

Did you want to roll the updates into this PR or wait for a future one? I'm fine either way!

alexcrichton commented on brotli-sys/src/lib.rs:
```diff
@@ -12,197 +12,81 @@ pub type __enum_ty = libc::c_int;
-pub type BrotliState = BrotliStateStruct;
-
-#[repr(C)]
-pub struct BrotliStateStruct {
```
Oh man it's so nice to have have to have this massive struct!

aidanhs commented:
Yeah I guess better to make the breaking changes sooner rather than later when it becomes more widely used. I'll put the updates into this PR.

aidanhs commented:
I think it's worth reviewing and considering this PR independently actually. Some benchmarking I did shows that this is noticeably slower, between 20 and 50%.

Here's a somewhat stripped version of the script I'm using to test:
```
extern crate brotli2;
extern crate rand;

use std::io::{Read, Write};
use brotli2::read;
use brotli2::write;
use brotli2::stream::{CompressParams, compress_buf, decompress_buf};
use rand::Rng;
use rand::SeedableRng;

// Used in functions as temporary storage space before producing a vec
const BIGBUF_SIZE: usize = 120*1024*1024;
static mut BIGBUF: [u8; BIGBUF_SIZE] = [0; BIGBUF_SIZE];

fn main() {
    let v1 = vec![1; 1024];
    let v2 = vec![44; 10*1024*1024];
    let mut params = CompressParams::new();
    params.mode(brotli2::stream::CompressMode::Text).quality(6);
    let params = &params;

    fn bufencode(data: &[u8], params: &CompressParams) -> Vec<u8> {
        let bufref = &mut unsafe { &mut BIGBUF[..] };
        let n = compress_buf(params, data, bufref).unwrap();
        assert!(n > 0 && n <= BIGBUF_SIZE && n == bufref.len());
        bufref.to_owned()
    }
    fn bufdecode(data: &[u8]) -> Vec<u8> {
        let bufref = &mut unsafe { &mut BIGBUF[..] };
        let s = decompress_buf(data, bufref).unwrap();
        assert!(s == brotli2::stream::Status::Finished);
        bufref.to_owned()
    }
    fn ioreadencode(data: &[u8], params: &CompressParams) -> Vec<u8> {
        let mut buf = vec![];
        read::BrotliEncoder::from_params(data, params).read_to_end(&mut buf).unwrap();
        assert!(buf.len() > 0);
        buf
    }
    fn ioreaddecode (data: &[u8]) -> Vec<u8> {
        let mut buf = vec![];
        read::BrotliDecoder::new(data).read_to_end(&mut buf).unwrap();
        buf
    }
    fn iowriteencode(data: &[u8], params: &CompressParams) -> Vec<u8> {
        let mut enc = write::BrotliEncoder::from_params(vec![], params);
        enc.write_all(data).unwrap();
        enc.finish().unwrap()
    }
    fn iowritedecode(data: &[u8]) -> Vec<u8> {
        let mut dec = write::BrotliDecoder::new(vec![]);
        dec.write_all(data).unwrap();
        dec.finish().unwrap()
    }

    let check = |data: &[u8]| {
        let c1 = &bufencode(data, params);
        let _ = &ioreadencode(data, params);
        let _ = &iowriteencode(data, params);
        assert!(data == &*bufdecode(c1));
        assert!(data == &*ioreaddecode(c1));
        assert!(data == &*iowritedecode(c1));
    };

    let mut rng = rand::XorShiftRng::from_seed([1, 4, 55, 98]);
    for _ in 0..5 {
        let rnum: usize = rng.gen_range(1, 100*1024*1024);
        let mut buf = vec![0; rnum];
        rng.fill_bytes(&mut buf);
        check(&buf)
    }
}
```

With 0.2.3:
```
real    0m17.171s
user    0m16.200s
sys     0m0.968s
```
With this PR:
```
real    0m21.711s
user    0m20.652s
sys     0m1.056s
```

I've not really changed anything other than doing the upgrade. Doing a very brief test with the `bro` tool doesn't show a similar slowdown.

alexcrichton commented:
So everything in this branch looks good to me, but the benchmark you posted seems worrisome! Is that an artifact of using the deprecated APIs? Does using the "right" apis in https://github.com/alexcrichton/brotli2-rs/pull/11 fix the speed loss?

If not, mind profiling to see where the time is spent? (e.g. in brotli vs in Rust)

aidanhs commented:
Reported at google/brotli#550

alexcrichton commented:
Awesome, thanks for investigating! Does that mean we're still good for upgrading, maybe just with known caveats?

aidanhs commented:
Turns out the delta between the last good version and v0.6.0 (where I was going to upgrade to) isn't that big (https://github.com/google/brotli/compare/cdca91b~1...v0.6.0), so I've just picked that last good commit.

Perf differences (measured by the example I added in the pre-upgrade commit, so that we can track over time) are about 10% slower for the new version, but it looks like there was a commit implementing better compression since the version we use so maybe that's expected.

Also, `decompress_buf` only returns true or false so there's no point in it returning a return status (it now returns a length, which is already available via `.len` I guess). I changed this in the pre-upgrade commit so the example can remain unchanged.

aidanhs commented:
Oh, should I add the example to the .travis.yml since it's also sort of a test? It takes about 12s.

alexcrichton commented:
Ok changes sound great to me! I'm gonna go ahead and merge but yeah feel free to throw in running the example to travis, 12s isn't so bad!
